### PR TITLE
[RISCV] Begin moving post-isel vector peepholes to a MF pass

### PR DESCRIPTION
**Pull request by lukel97**
We currently have three postprocess peephole optimisations for vector pseudos:

1) Masked pseudo with all ones mask -> unmasked pseudo
2) Merge vmerge pseudo into operand pseudo's mask
3) vmerge pseudo with all ones mask -> vmv.v.v pseudo

This patch aims to move these peepholes out of SelectionDAG and into a separate RISCVFoldMasks MachineFunction pass.

There are a few motivations for doing this:

* The current SelectionDAG implementation operates on MachineSDNodes, which are essentially MachineInstrs but require a bunch of logic to reason about chain and glue operands. The RISCVII::has*Op helper functions also don't exactly line up with the SDNode operands. Mutating these pseudos and their operands in place becomes a good bit easier at the MachineInstr level. For example, we would no longer need to check for cycles in the DAG during performCombineVMergeAndVOps.

* Although it's further down the line, moving this code out of SelectionDAG allows it to be reused by GlobalISel later on.

* In performCombineVMergeAndVOps, it may be possible to commute the operands to enable folding in more cases (see test/CodeGen/RISCV/rvv/vmadd-vp.ll). There is existing machinery to commute operands in TII::commuteInstruction, but it's implemented on MachineInstrs.

The pass runs straight after ISel, before any of the other machine SSA optimization passes run. This is so that dead-mi-elimination can mop up any vmsets that are no longer used (but if preferred we could try and erase them from inside RISCVFoldMasks itself). This also means that these peepholes are no longer run at codegen -O0, so this patch isn't strictly NFC.

Only the performVMergeToVMv peephole is refactored in this patch, the remaining two would be implemented later. And as noted by @preames, it should be possible to move doPeepholeSExtW out of SelectionDAG as well.

### CODE REVIEW

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVFoldMasks.cpp [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  if (MergeReg != RISCV::NoRegister && TRI->lookThruCopyLike(MergeReg, MRI) !=
+                                           TRI->lookThruCopyLike(FalseReg, MRI))
+    return false;
+
+  assert(MI.getOperand(4).isReg() && MI.getOperand(4).getReg() == RISCV::V0);
+  if (!isAllOnesMask(V0Def))
+    return false;
+
+  MI.setDesc(TII->get(NewOpc));
+  MI.removeOperand(2); // False operand
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:133]: I think you might have a bug here.  I think you want to be dropping the merge operand, not the false operand.  In the case where merge == false, it doesn't matter.  But in the case where merge is undef, I think you need to keep the false operand, not the merge operand.

**Review by preames — APPROVED** [review]
LGTM w/one fix.

File: llvm/lib/Target/RISCV/RISCVFoldMasks.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+  case RISCV::PseudoVMSET_M_B64:
+    return true;
+  default:
+    return false;
+  }
+}
+
+static bool isVMerge(MachineInstr &MI) {
+  unsigned Opc = MI.getOpcode();
+  return Opc == RISCV::PseudoVMERGE_VVM_MF8 ||
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:93]: Rebase is needed after #70637

File: llvm/lib/Target/RISCV/RISCVFoldMasks.cpp (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVFoldMasks.cpp]: Yup, thanks for catching this. Should be fixed now, and added a test case

**Pull request merged**